safehttp/plugins/xsrf/xsrf.go

```diff
@@ -0,0 +1,172 @@
+	// FMEnforceMode TODO
+	FMEnforceMode = "enforce"
+	// FMReportMode TODO
+	FMReportMode = "report"
+	// FMDisabledMode TODO
+	FMDisabledMode = "disable"
+	// TokenKey is the form key used when sending the token as part of POST
+	// request
+	TokenKey = "xsrf-token"
+)
+
+// StorageService is an interface the framework users need to implement. This
+// will contain information about users of the web application, including their
+// IDs, needed in generating the XSRF token.
+type StorageService interface {
+	// TODO(@mihalimara22): add the parameters that the storage service needs in
+	// order to determine the user ID
+	GetUserID() (string, error)
+}
+
+// Plugin implements XSRF protection.
+type Plugin struct {
+	// TODO(@mihalimara22): decide whether  we want to allow the user to set a
+	// Fetch Metadata policy and how the report mode is going to be handled
+	appKey       string
+	s            StorageService
+	fmPolicyMode string
+	allowedCORS  map[string]bool
+}
+
+// NewPlugin creates a new XSRF plugin with safe defaults
+func NewPlugin(appKey string, s StorageService, domains ...string) *Plugin {
+	// TODO(@mihalimara22): make endpoints a variadic of string literals
+	ad := map[string]bool{}
+	for _, p := range domains {
+		ad[p] = true
+	}
+	return &Plugin{
+		appKey:       appKey,
+		s:            s,
+		fmPolicyMode: FMEnforceMode,
+		allowedCORS:  ad,
+	}
+}
+
+// SetFetchMetadataPolicyMode allows setting the Fetch Metadata policy, enforced
+// by default to one of the other options. Setting it to report will allow
+// requests that violate the policy to pass but will log violation. Setting it
+// to disable
+func (p *Plugin) SetFetchMetadataPolicyMode(mode string) error {
+	switch mode {
+	case FMEnforceMode, FMDisabledMode, FMReportMode:
+		p.fmPolicyMode = mode
+	default:
+		return errors.New("invalid Fetch Metadata policy")
+	}
+	return nil
+}
+
+// GenerateToken generates a cryptographically safe XSRF token using the user ID
+// and the path.
+func (p *Plugin) GenerateToken(host string) (string, error) {
+	userID, err := p.s.GetUserID()
+	if err != nil {
+		return "", fmt.Errorf("token generation failed: %v", err)
+	}
+	token := xsrftoken.Generate(p.appKey, userID, host)
+	return token, nil
+}
+
+// validateToken validates the XSRF token. This should be present in all
+// requests as the value of form parameter xsrf-token.
+func (p *Plugin) validateToken(ir *safehttp.IncomingRequest) (safehttp.StatusCode, bool) {
+	userID, err := p.s.GetUserID()
+	if err != nil {
+		return safehttp.Status401Unauthorized, false
+	}
+	// TODO(@mihalimara22): add multipart support as well
+	f, err := ir.PostForm()
+	if err != nil {
+		return safehttp.Status400BadRequest, false
+	}
+	token := f.String(TokenKey, "")
+	if f.Err() != nil || token == "" {
+		return safehttp.Status403Forbidden, false
+	}
+	ok := xsrftoken.Valid(token, p.appKey, userID, ir.Header.Get("Host"))
+	if !ok {
+		return safehttp.Status403Forbidden, false
+	}
+	return 0, true
```

Comment:
Shouldn't this be `return safehttp.StatusOK, true`?
Same in `applyPolicy`

Reply:
I also considered that option but wasn't sure. The thing is, if validation succeeds we will move on to the handler or another plugin so there is no way of knowing the status code is going to be 200. Therefore, I thought it's a better option (for now) not to return a status code unless something wrong occurs and we have to return immediately. But idk what is best.

Reply:
I see your point about not returning `200`. I would suggest that `0` is returned on success and that the `bool` is removed, it feels unnecessary.